indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,685 @@
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Queue for {@link SegmentAllocateRequest}s.
+ */
+@ManageLifecycle
+public class SegmentAllocationQueue
+{
+  private static final Logger log = new Logger(SegmentAllocationQueue.class);
+
+  private final long maxWaitTimeMillis;
+  private final boolean enabled;
+
+  private final TaskLockbox taskLockbox;
+  private final ScheduledExecutorService executor;
+  private final IndexerMetadataStorageCoordinator metadataStorage;
+  private final DruidLeaderSelector leaderSelector;
+  private final ServiceEmitter emitter;
+
+  private final ConcurrentHashMap<AllocateRequestKey, AllocateRequestBatch> keyToBatch = new ConcurrentHashMap<>();
+  private final Deque<AllocateRequestBatch> processingQueue = new ConcurrentLinkedDeque<>();
+
+  @Inject
+  public SegmentAllocationQueue(
+      TaskLockbox taskLockbox,
+      TaskLockConfig taskLockConfig,
+      IndexerMetadataStorageCoordinator metadataStorage,
+      @IndexingService DruidLeaderSelector leaderSelector,
+      ServiceEmitter emitter
+  )
+  {
+    this.emitter = emitter;
+    this.taskLockbox = taskLockbox;
+    this.metadataStorage = metadataStorage;
+    this.leaderSelector = leaderSelector;
+    this.maxWaitTimeMillis = taskLockConfig.getBatchAllocationMaxWaitTime();
+    this.enabled = taskLockConfig.isBatchSegmentAllocation();
+
+    this.executor = ScheduledExecutors.fixed(1, "SegmentAllocQueue-%s");
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    log.info("Starting queue.");
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    log.info("Stopping queue.");
+    executor.shutdownNow();
+  }
+
+  public boolean isEnabled()
+  {
+    return enabled;
+  }
+
+  private void scheduleQueuePoll(long delay)
+  {
+    executor.schedule(this::processBatchesDue, delay, TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Queues a SegmentAllocateRequest. The returned future may complete successfully
+   * with a non-null value or with a non-null value.
+   */
+  public Future<SegmentIdWithShardSpec> add(SegmentAllocateRequest request)
+  {
+    if (!leaderSelector.isLeader()) {
+      throw new ISE("Cannot allocate segment if not leader.");
+    } else if (!isEnabled()) {
+      throw new ISE("Batched segment allocation is disabled.");
+    }
+
+    final AllocateRequestKey requestKey = new AllocateRequestKey(request, false);
+    final AtomicReference<Future<SegmentIdWithShardSpec>> requestFuture = new AtomicReference<>();
+
+    keyToBatch.compute(requestKey, (key, existingBatch) -> {
+      AllocateRequestBatch computedBatch = existingBatch;
+      if (computedBatch == null) {
+        computedBatch = new AllocateRequestBatch(key);
+        computedBatch.resetQueueTime();
+        processingQueue.offer(computedBatch);
```

Comment:
what if `offer` returns `false`?

Reply:
Thanks, will handle this. If the batch couldn't be added to the queue (due to the queue being full), I suppose we don't have a choice but to just complete the request with a `null` return value.

Reply:
Updated.